# ENCH 407 - Assignment 2 - Winter 2022

**GROUP #:** *Fill your group number here*

**Group members:** *Fill your name(s) here*

*Due date: February 28, 2022: 5 PM*

**MAKE SURE TO SHOW ALL CODING YOU USED IN YOUR WORK IN THE WORKSHEET BELOW AND MAKE SURE YOUR CODE RUNS!**

In a chemical engineering plant, a series of reactors and separators are present as depicted below. We are concerned with the water management within this plant, and want to know what the concentrations of water throughout the plant are. 

<img><img src='Assignment2_VesselNetwork.png'></img>

There are three locations in which water is entering the plant (into Vessel 1, Vessel 2 and Vessel 3) with a variable concentration as a function of time. Out of two locations (Vessel 3 and Vessel 4), water leaves the plant. Part of the aqueous process fluid from Vessel 4 is recycled within the process. You may assume the system is at steady state at all times when doing your calculations. 

## Part 1 - Defining the system of equations (15 Points)

* Using the conservation equation $\left(\sum (Qc)_{in}+ \sum(Qc)_{out} = 0\right)$ for all of the vessels, write the system of equations required to find the concentrations $c$ in all of the vessels. Define your matrix, the vector of unknowns and vector of constants for the universal situation in which none of the values for $Q_{ij}$ and the concentrations are given. Keep in mind that the sign of ingoing and outgoing streams should be different for the conservation equation to hold. 


## Answer cell for part 1

Write your equations here:
$$
\begin{equation}
\begin{bmatrix}
Q_{01}*c_{1,in}+Q_{41}*c_{4}-Q_{12}*c_{1}=0 \\
Q_{02}*c_{2,in} +Q_{12}*c_{1}-Q_{23}*c_{2}-Q_{24}*c_{2}=0 \\
Q_{23}*c_{2} +Q_{03}*c_{3,in}-Q_{30}*c_{3}-Q_{34}*c_{3}=0  \\
Q_{34}*c_{3} +Q_{24}*c_{2}-Q_{40}*c_{4}-Q_{41}*c_{4}=0 
\end{bmatrix}
\end{equation}
$$

Fill the general matrix $[A]$ below by substituting $a_{ij}$ with their actual values:

$$
\begin{equation}

[A] = 

 \begin{bmatrix}
  -Q_{12} & 0 & 0 & Q_{41} \\
  Q_{12} & -Q_{23}-Q_{24} & 0 & 0 \\
  0 & Q_{23} & -Q_{30}-Q_{34} & 0 \\
  0 & Q_{24} & Q_{34} & -Q_{40}-Q_{41}
 \end{bmatrix}

 \end{equation}
$$

Fill the vector of unknowns by substituting your values on the dots:


$$
\begin{equation}
{x} = 

\begin{bmatrix}
c_{1} \\
c_{2}  \\
c_{3}  \\
c_{4}  \\
\end{bmatrix}
\end{equation}
$$

Fill the general vector of constants below by substituting your values on the dots: 

$$
\begin{equation}
{b} = 

\begin{bmatrix}
-Q_{01}*c_{1,in} \\
-Q_{02}*c_{2,in} \\
-Q_{03}*c_{3,in} \\
0 \\
\end{bmatrix}
\end{equation}
$$

# Part 2 - Solving the system of equations (total 50 points)

For the plant we are concerned with, the volumentric flow rates $Q_{ij}$ between vessels are fixed and have the values as in per the table below. 

|Stream| Flow rate ($\mathrm{m}^3\mathrm{/hr}$) |
|---|---|
|$Q_{01}$ | 5.0 | 
|$Q_{02}$ | 15.0 | 
|$Q_{03}$ | 10.0 |
|$Q_{12}$ | 15.0 |
|$Q_{23}$ | 10.0 | 
|$Q_{24}$ | 20.0 |
|$Q_{30}$ | 5.0 |
|$Q_{34}$ | 15.0 |
|$Q_{40}$ | 25.0 |
|$Q_{41}$ | 10.0 |

** What numerical methods we have discussed would be useful for solving this system of equations? Which method do you consider the most suitable considering the inlet concentrations might fluctuate? Why do you choose this method? Describe the advantages and disadvantages of the method you deem most suitable.  (15 points)
- LU decpmposition is probably the best method in this case since it does not depend on the constant vector and that vector can change by time. In this case, the concentration is not fixed so in case of using other method, whenever changing the constant it requires to do the whole process again which can use up a lot of computational space. However, with the LU decomposition, it requires less time to compute the answer.  

** Is pivoting required when the system has the given values? Why or why not? (5 points)**
- Pivoting is rquired to solve this matrix as in the second column and third column it is not well-organized. the pivot element can be NAN when we do the problem because of the zeros. 
** Write the function for this method to find the concentrations of species $c_i$ in the four vessels in the case that $c_{1,in} = 900$ kg/m $^3$, $c_{2,in}= 950$ kg/m $^3$ and $c_{3,in} = 800$ kg/m $^3$. (20 points)

** Next, use your code to calculate the concentration in all of the vessels in the case that the inlet concentrations are disturbed by a process event upstream from this plant section and the inlet concentrations are $c_{1,in} = 800$ kg/m $^3$, $c_{2,in}= 950$ kg/m $^3$ and $c_{3,in} = 1000$ kg/m $^3$. (10 points)**

In [30]:
# # Use this cell to define the function for the method you are choosing to use to solve this system of equations

# import numpy as np 

# def LUdecomp(a):                        # This function takes matrix a, and returns a matrix of the same size that contains both L and U
#     n = len(a)
#     for k in range(n-1):
#         for i in range (k+1,n):            #for pivoting element in k, we want to eliminate all values below from row i to n
#             if np.abs(a[i,k]) > np.abs(a[k,k]):     # check if we need pivoting
#                 a[[k,i]] = a[[i,k]]                 # pivot matrix a
#                 b[[k,i]] = b[[i,k]]   
#             if a[i,k] != 0:
#                 fac = a[i,k]/a[k,k]     #pivoting factor, where a[k,k] is pivoting element
#                 a[i,k+1:n] = a[i,k+1:n] - fac*a[k,k+1:n]
#                 a[i,k] = fac 
#                 # print(a)           # Store the pivoting factor in the bottom triangular of matrix a
#     return a

# def LUsolve(a,b):                   # a should be the LU matrix as input here 
#     n = len(a)
#     for k in range(1,n):
#         b[k] = b[k] - np.dot(a[k,0:k],b[0:k])       # Changing B to D
#     for k in range(n-1,-1,-1):
#         b[k] = (b[k]-np.dot(a[k,k+1:n],b[k+1:n]))/a[k,k]
#     return b

In [31]:
# Use this cell to calculate the required concentrations 

# a = np.array([[-15, 0, 0,10],[15, -30, 0, 0],[0, 10, -20, 0],[0, 20, 15, -35]])      # For this matrix we require pivoting, which is not written into LU-decomp
# b = np.array([-4500,-14250,-8000, 0],float)

# a = LUdecomp(a)
# print(a)
# x = LUsolve(a,b)

# print(x)

# a = np.array([[-15, 0, 0,10],[15, -30, 0, 0],[0, 10, -20, 0],[0, 20, 15, -35]])      # For this matrix we require pivoting, which is not written into LU-decomp
# b = np.array([-4500,-14250,-8000, 0],float)

# a = LUdecomp(a)
# print(a)
# x = LUsolve(a,b)
# print(x)

In [36]:
def gaussPivot(a,b):
    n = len(b)
    for k in range(n-1):
        for i in range(k+1,n):
            if a[i,k] == 0:
                continue                # continue makes python skip the rest of the for loop
            if np.abs(a[i,k]) > np.abs(a[k,k]):     # check if we need pivoting
                a[[k,i]] = a[[i,k]]                 # pivot matrix a
                b[[k,i]] = b[[i,k]]                 # pivot vector b
            # Starting our elimination
            fac = a[i,k]/a[k,k]
            a[i,k+1:n] = a[i,k+1:n] -fac*a[k,k+1:n]
            b[i] = b[i] - fac*b[k]
    for k in range(n-1,-1,-1):
        b[k] = (b[k]-np.dot(a[k,k+1:n],b[k+1:n]))/a[k,k]
    return b

In [37]:
a = np.array([[-15, 0, 0,10],[15, -30, 0, 0],[0, 10, -20, 0],[0, 20, 15, -35]])      # For this matrix we require pivoting, which is not written into LU-decomp
b = np.array([-4500,-14250,-8000, 0],float)

print('a =', a)

x = gaussPivot(a,b)
print(x)

a = np.array([[-15, 0, 0,10],[15, -30, 0, 0],[0, 10, -20, 0],[0, 20, 15, -35]])      # For this matrix we require pivoting, which is not written into LU-decomp
b = np.array([-4000,-14250,-10000, 0],float)

print('a =', a)

x = gaussPivot(a,b)
print(x)

a = [[-15   0   0  10]
 [ 15 -30   0   0]
 [  0  10 -20   0]
 [  0  20  15 -35]]
[918.33333333 934.16666667 851.625      927.5       ]
a = [[-15   0   0  10]
 [ 15 -30   0   0]
 [  0  10 -20   0]
 [  0  20  15 -35]]
[912.77777778 931.38888889 949.54166667 969.16666667]


## Answer cell for Part 2

* For solving this system of equations, we will use **YOUR ANSWER HERE** method, because ** YOUR EXPLANATION HERE **.

* Pivoting is ** REQUIRED / NOT REQUIRED (CHOOSE YOUR ANSWER) ** for this system because ** YOUR EXPLANATION HERE **.

* With this method, we find the following concentrations in the case that $c_{1,in} = 900$ kg/m $^3$, $c_{2,in}= 950$ kg/m $^3$ and $c_{3,in} = 800$ kg/m $^3$ :
$$
\begin{equation}
\begin{matrix}
c_1 =  [UNIT] \\
c_2 =  [UNIT] \\
c_3 =  [UNIT] \\
c_4 =  [UNIT] \\
\end{matrix}
\end{equation}
$$
*  With this method, we find the following concentrations in the case that $c_{1,in} = 800$ kg/m $^3$, $c_{2,in}= 950$ kg/m $^3$ and $c_{3,in} = 1000$ kg/m $^3$ :

$$
\begin{equation}
\begin{matrix}
c_1 =  [UNIT] \\
c_2 =  [UNIT] \\
c_3 =  [UNIT] \\
c_4 =  [UNIT] \\
\end{matrix}
\end{equation}
$$

# Part 3 - Iterative methods (total 35 points)

* Write a Python code for the Gauss-Seidel method to iteratively solve the system of equations to find the concentrations of species $c_i$ in the four vessels in the case that $c_{1,in} = 800$ kg/m $^3$, $c_{2,in}= 950$ kg/m $^3$ and $c_{3,in} = 1000$ kg/m $^3$, with an $\epsilon_s = 10^{-6}$ and initial guesses for all concentrations of 1000. Check your solution by comparing it to the solution found under Part 2. (15 points)
* How many iterations are required to find the solution using this method? (5 points)
* Compare the current method with your previously used method and comment on the differences in both methods. (15 points)

In [34]:
# Use this cell to write your Gauss Seidel method

import numpy as np

def GaussSeidel(a,b,epsilon,max_iter):
    n = len(b)
    x = np.array([1000,1000,1000,1000], float)                    # Initial guesses for x1, x2 ... xn, x is a vector with n values
    eps_a = 100*np.ones_like(b)                 # Initial epsilon values        eps_a is vector with n values
    for k in range(max_iter):
        eps_max = np.max(np.abs(eps_a))        # Finding the largest error we have after the iteration 
        if eps_max < epsilon:
            print('Found a solution after', k, 'iterations')
            return x
        else:
            for i in range(n):              # For n equations
                sum = 0.0 
                for j in range(n):
                    if (j!=i):
                        sum += a[i,j]*x[j]
                        x_temp = x[i]
                x[i] = (b[i]-sum)/a[i,i]        # x1 = (b1 - a12x2 - a13x3)/a11, x2 = (b2 - a21x1 - a23x3)/a22 
                eps_a[i] = (x[i]-x_temp)/x[i]*100
    print('Exceeded maximum number of iterations, no solution found') 
    return None

In [35]:
# Use this cell to calculate the vessel concentrations using your function

a = np.array([[-15, 0, 0,10],[15, -30, 0, 0],[0, 10, -20, 0],[0, 20, 15, -35]])      # For this matrix we require pivoting, which is not written into LU-decomp
b = np.array([-4000,-14250,-10000, 0],float)

x = GaussSeidel(a,b,10**(-6),100)
print(x)

Found a solution after 14 iterations
[891.935485   920.9677425  960.48387125 937.90322625]


## Answer cell for Part 3

* Our Gauss Seidel method found the solution for the concentrations after 14 iterations

* With this method, we find the following concentrations:
$$
\begin{equation}
\begin{matrix}
c_1 =  891.935 \\
c_2 =  920.968 \\
c_3 =  960.484 \\
c_4 =  937.903 
\end{matrix}
\end{equation}
$$
* When comparing this method to our previous method, we can observe that the Gauss-Seidel does not require any pivoting to compute the final answer. However, it requires an initial guess to solve the problem. If we choose an initial guess that is far away from the final answer, it takes up a lot of time and computational space to find the answer. 